core/src/scene/sceneLayer.cpp

```diff
@@ -1,14 +1,61 @@
 #include "sceneLayer.h"
 
+#include <algorithm>
+
 namespace Tangram {
 
-uint8_t SceneLayer::s_layerCount = 0;
+SceneLayer::SceneLayer(std::string _name, Filter _filter, std::vector<DrawRule> _rules, std::vector<SceneLayer> _sublayers) :
+    m_filter(_filter), m_name(_name), m_rules(_rules), m_sublayers(_sublayers), m_depth(0) {
+
+    // Rules must be sorted to merge correctly
+    std::sort(m_rules.begin(), m_rules.end());
+
+    // m_depth is one more than the maximum depth of any sublayer
+    for (auto& sublayer : m_sublayers) {
+        m_depth = sublayer.m_depth > m_depth ? sublayer.m_depth : m_depth;
+    }
+    m_depth++;
+
+    // Sublayers must be sorted by the depth of their deepest leaf in order to correctly traverse them in match()
+    std::sort(m_sublayers.begin(), m_sublayers.end(), [&](const SceneLayer& a, const SceneLayer& b) {
+        return a.m_depth > b.m_depth;
+    });
+
+}
+
+void SceneLayer::match(const Feature& _feat, const Context& _ctx, std::vector<DrawRule>& _matches) const {
```

Comment:
Why not return the vector of DrawRules here, instead of pass by reference and swapping?
(I am thinking in reference to let compiler do its job, based on the explanation given by the llvm guy :D).